### PR TITLE
Dynamic plugins list in the website

### PR DESCRIPTION
**Pull request by ahmetb**
This relies on Netlify's serverless functions capability to render
plugin count (in the homepage) and plugin list (/docs/plugins) in
the website!

SIG Docs TL has cleared the usage of Functions feature here: 
https://groups.google.com/g/kubernetes-sig-docs/c/cXqQm60tGs8

The serverless function is called "api" and it's accessible on path
/.netlify/functions/api/* and has two subpaths that return JSON responses:

1. */pluginCount
2. */plugins

This serverless function is built by placing a binary in /site/functions
named "api" and the build instruction is in 'netlify.toml' file.

Since we get plugin list by calling GitHub API, by default we run without
an authentication token. However after this is merged, I will go and add a
permissionless $GITHUB_ACCESS_TOKEN that I have created from my account.

Unauthenticated GitHub calls are limited to 60 per IP, but since I'll add a
personal access token, that should not be an issue. When debugging locally,
60 is also fine. (deploy previews for PRs from non-maintainers will not have
an access token available, so they will make unauthenticated calls to GH API.)

The responses from the dynamic functions are actually cached in Netlify's
CDN and that helps with the rate-limiting as well. I have currently set the
CDN cache duration to 1 hour for each response.

Please take a look at the previews and let me know if there is anything that
doesn't look good.

I might delete `generate-plugin-overview` tool and update the current
plugins.md in a followup PR.

/assign @chriskim06
/assign @corneliusweig

### CODE REVIEW

**Comment by ahmetb** [comment]
/hold
Not entirely ready for review, requires some setup. I'll ping again later.

**Comment by ahmetb** [comment]
/hold cancel
Ready to preview at https://deploy-preview-658--kubernetes-sigs-krew.netlify.app/.
Feel free to add commits to my branch.

File: site/functions/server/main.go
Comment (corneliusweig) [inline site/functions/server/main.go:116]: Let's not do this and save a few bits on the wire.
If we need to look at this, it will most certainly be in a browser console and they print json objects always the same.
```suggestion
```

File: site/functions/server/main.go
Comment (corneliusweig) [inline site/functions/server/main.go:180]: `ctx` should be passed in from `req.Context()`

File: site/functions/server/main.go
Comment (corneliusweig) [inline site/functions/server/main.go:195]:
```suggestion
		go func() {
```
`j` is unused?

File: site/functions/server/main.go
Comment (corneliusweig) [inline site/functions/server/main.go:193]: nit: urlFetchBatchSize -> pluginFetchParallelism
```suggestion
	for i := 0; i < pluginFetchParallelism; i++ {
```

File: site/functions/server/main.go
Comment (corneliusweig) [inline site/functions/server/main.go:191]: Maybe use errgroup instead and avoid the `context.WithCancel`?

File: site/functions/server/main.go
Comment (corneliusweig) [inline site/functions/server/main.go:215]: When using errgroup, this could be simplified to

```go
for url := range queue {
  p, err := readPlugin(url)
  if err != nil {
    return err
  }
  mu.Lock()
  out = append(out, p)
  mu.Unlock()
}
```

File: site/functions/server/main.go
Comment (corneliusweig) [inline site/functions/server/main.go:237]: nit: run gofmt :)

File: site/functions/server/main.go
Comment (corneliusweig) [inline site/functions/server/main.go:241]: nit:
```suggestion
	defer resp.Body.Close()
```

From the spec:
> // The http Client and Transport guarantee that Body is always
	// non-nil, even on responses without a body or responses with
	// a zero-length body. It is the caller's responsibility to
	// close Body.

File: site/functions/server/main.go
Comment (corneliusweig) [inline site/functions/server/main.go:242]: nit: move `v` right before `yaml.Unmarshal`

File: site/functions/server/main.go
Comment (corneliusweig) [inline site/functions/server/main.go:280]: The comment reads like this should be commented out?

Anyway, I think it'd be much nicer like that, then no commenting is necessary:
```go
if *port > -1 {
   ...
}
```

**Review by corneliusweig — APPROVED** [review]
This is really nice work! A few nits, but otherwise this is perfect stuff.

Very cool that you found out about netlify functions. This will save us so much pointless work.

---

I agree that `generate-plugin-overview` should be deleted asap as soon as this is merged.

File: site/layouts/partials/footer.html
Comment (ahmetb) [reply site/layouts/partials/footer.html:57]: Might do `<error>` :D

File: site/functions/server/main.go
Comment (ahmetb) [reply site/functions/server/main.go:280]: It can be commented out but I left it on just in case. You're right.

**Comment by ahmetb** [comment]
@corneliusweig I think it would be better if you do a pass with your commits on the frontend pieces. My frontend skills are weak at best. :)

File: site/functions/server/main.go
Comment (ahmetb) [reply site/functions/server/main.go:215]: might use errgroup, but I don't see a way to limit parallelism with errgroup.
feel free to add a patch.

**Comment by corneliusweig** [comment]
@ahmetb I added some commits but had to create a PR to your repo: https://github.com/ahmetb/krew/pull/2
Please have a look

**Comment by ahmetb** [comment]
🤔 I see "Allow edits and access to secrets by maintainers" so you should be able to push to my branch. Have you tried over ssh (git://) ?

File: site/functions/server/main.go
Comment (chriskim06) [inline site/functions/server/main.go:170]: Wouldn't it be fine to just pass `v.Spec.Homepage` here? Then plugins from gitlab or somewhere not in the `knownHomePages` map could still be linked

Comment (chriskim06) [reply site/functions/server/main.go:170]: wait nvm I just realized there's also Homepage being passed here and this is just for the github stars link

**Comment by corneliusweig** [comment]
So I tried again various approaches---nothing will update this PR. However, now there's also a new branch in krew, called `netlify-dynamic-content` :(.

So I think it's easiest if you merge my PR into your PR :roll_eyes:

File: site/functions/go.mod
Comment (ahmetb) [inline site/functions/go.mod:13]: this is a bit worrisome to pin the site to a particular version of `master`. (should be ok as long as we keep the types/consts backwards compatible)

this `replace` statement works fine locally, but once I uncommented the `replace` statement and deploy, it fails to build on Netlify sadly. not sure why.

File: site/functions/server/main.go
Comment (ahmetb) [inline site/functions/server/main.go:251]: @corneliusweig I wonder if we should move this to somewhere like a `var` declaration at the top. I suspect it'll get updated frequently. :D

File: site/layouts/partials/footer.html
Comment (ahmetb) [inline site/layouts/partials/footer.html:47]: @corneliusweig should we move these to a .js file at this point?

File: site/layouts/partials/footer.html
Comment (ahmetb) [inline site/layouts/partials/footer.html:66]: I must admit I don't fully understand how imports work in the browser so @corneliusweig you’ll probably be on point for maintaining these :D I barely can get the vanilla js working.

File: site/functions/go.mod
Comment (corneliusweig) [reply site/functions/go.mod:13]: Yes, it will be annoying if we forget about it. I don't see a better alternative though.

File: site/functions/server/main.go
Comment (corneliusweig) [reply site/functions/server/main.go:251]: I wonder if we should even go one step further and extract this configuration into non-code. In `krew-index-tracker`, I also need this list and it is error-prone to redeploy every time this changes.

Maybe we can keep this map in a very dumb format in a text file inside `krew-index` or `krew`. Something like
```txt
https://krew.sigs.k8s.io/ kubernetes-sigs/krew
https://sigs.k8s.io/krew kubernetes-sigs/krew
...
```

File: site/layouts/partials/footer.html
Comment (corneliusweig) [reply site/layouts/partials/footer.html:47]: Yes, that makes sense to me. We don't have to do it in this PR though. Maybe open a new issue? [[main reason: collaboration on the shared PR seems to be not straightforward]]

File: site/layouts/partials/footer.html
Comment (corneliusweig) [reply site/layouts/partials/footer.html:66]: Yeah, I'm not too worried about it ((support stats)[https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import]).

File: site/functions/server/main.go
Comment (corneliusweig) [inline site/functions/server/main.go:256]: Running the server locally with `hugo` did not work for me, I always got `not found`, because the site was trying to query the server on port 1313 (hugo) instead of 8080. Can you write down the instructions in a step-by-step manner?

File: site/layouts/partials/footer.html
Comment (chriskim06) [inline site/layouts/partials/footer.html:57]: It seems like optional chaining is relatively [new](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining#Browser_compatibility) (or at least those version numbers seem a little high). Do either of you think this should be changed to the old way or is it ok?

Comment (corneliusweig) [reply site/layouts/partials/footer.html:57]: The question immediately becomes: how far back do we want to be backwards compatible.

This case is trivial, but `import` and `async/await` is already more involved and the old style quickly becomes kind of ugly. We could warn page visitors by including a `<script nomodule ..>` script to inform the user that the page is not working as intended and to switch to a newer browser.

Comment (ahmetb) [reply site/layouts/partials/footer.html:57]: I think these are mostly fine. I tend to assume developers use latest browser versions most of the time. In that regard they're more savvy than average computer user.

File: site/functions/server/main.go
Comment (ahmetb) [reply site/functions/server/main.go]: pushed a ./site/functions/README.md . PTAL

File: site/layouts/partials/footer.html
Comment (ahmetb) [reply site/layouts/partials/footer.html:47]: ack

File: site/functions/server/main.go
Comment (ahmetb) [reply site/functions/server/main.go:251]: I'm -1 for a new file/format.
I think we can hardcode these still in a centralized source code file.
We can do a variable on top of main.go for now.

**Comment by ahmetb** [comment]
added two more commits:

1. 543f132 move from /docs/plugins to /plugins → this way plugin list isn't part of docs
1. fb0b188 add /plugins to navbar → show a "Plugins" item on the top navigation bar.

**Comment by corneliusweig** [comment]
Oh no! I'm so slow. Apologies 🙇‍♂️ 
/lgtm
/approve

**Pull request merged**
